### PR TITLE
Add an in-memory cache in `Trial` class

### PR DESCRIPTION
**Pull request by eukaryo**
## Motivation

make Optuna storage faster.

cf. [https://gist.github.com/c-bata/16d654471393f59e064482ffb8b52bf8](https://gist.github.com/c-bata/16d654471393f59e064482ffb8b52bf8)

> Optuna uses `CachedStorage`, a wrapper class of `BaseStorage` interface, since the API calls of `BaseStorage` interface tends to be expensive. However, the implementation of `CachedStorage` is too complex and it's actually not a wrapper of `BaseStorage` since we've introduced some private storage APIs for `CachedStorage` like [storage._check_and_set_param_distribution()](https://github.com/optuna/optuna/blob/master/optuna/storages/_cached_storage.py#L238-L240) and [storage._get_trials()](https://github.com/optuna/optuna/blob/master/optuna/storages/_cached_storage.py#L387-L389).
So I implemented a prototype of a new simple caching mechanism to remove `CachedStorage` from Optuna. The change is only about 50 lines, but it is more efficient in many situations than CachedStorage.

## Description of the changes

cf. [https://gist.github.com/c-bata/16d654471393f59e064482ffb8b52bf8](https://gist.github.com/c-bata/16d654471393f59e064482ffb8b52bf8)

### CODE REVIEW

File: optuna/study/study.py [hunk 1]
```diff
@@ -195,7 +197,7 @@ def direction(self) -> StudyDirection:
 
         return self.directions[0]
 
-    @property
+    @cached_property
```
Comment (c-bata) [inline optuna/study/study.py:200]: This cache logic is introduced in another pull request. So could you revert this change?

File: optuna/trial/_trial.py [hunk 2]
```diff
@@ -696,7 +702,7 @@ def system_attrs(self) -> Dict[str, Any]:
             A dictionary containing all system attributes.
         """
 
-        return copy.deepcopy(self.storage.get_trial_system_attrs(self._trial_id))
+        return self._cached_frozen_trial.system_attrs
```
Comment (c-bata) [inline optuna/trial/_trial.py:705]: Could you returns the `self._cached_frozen_trial.user_attrs` in `user_attrs` property method?

In terms of `system_attrs`, please keep the existing logic since it will be deprecated.

**Review by c-bata — CHANGES_REQUESTED** [review]

File: optuna/trial/_trial.py (hunk 2 above)
Comment (eukaryo) [reply optuna/trial/_trial.py]: I revised the code.
https://github.com/eukaryo/optuna/commit/08b951cbb4586af496db48bbf07b5bf4c5233b8c

File: optuna/study/study.py (hunk 1 above)
Comment (eukaryo) [reply optuna/study/study.py]: I have reverted the change.
https://github.com/optuna/optuna/pull/4240/commits/ee429a3937986549731bf96168632b1286b65f6b

**Review by amylase — CHANGES_REQUESTED** [review]

File: optuna/trial/_trial.py [hunk 3]
```diff
@@ -686,7 +687,7 @@ def user_attrs(self) -> Dict[str, Any]:
             A dictionary containing all user attributes.
         """
 
-        return copy.deepcopy(self.storage.get_trial_user_attrs(self._trial_id))
+        return self._cached_frozen_trial.user_attrs
```
[the remark this answers is not in the document]
Comment (eukaryo) [reply optuna/trial/_trial.py]: I revised the code. I hope the new code reflects your intentions correctly.
https://github.com/eukaryo/optuna/commit/27070560524902df9288cf05e5fd0b5e7ace1475

File: optuna/trial/_trial.py [hunk 4]
```diff
@@ -601,11 +602,13 @@ def _suggest(self, name: str, distribution: BaseDistribution) -> Any:
             param_value_in_internal_repr = distribution.to_internal_repr(param_value)
             storage.set_trial_param(trial_id, name, param_value_in_internal_repr, distribution)
 
+        self._cached_frozen_trial.distributions[name] = distribution
+        self._cached_frozen_trial.params[name] = param_value
```
[the remark this answers is not in the document]
Comment (eukaryo) [reply optuna/trial/_trial.py]: I revised the code.
https://github.com/eukaryo/optuna/commit/55a6ee2d9af91176b3a09619640fae2c2953bdca

File: optuna/trial/_trial.py [hunk 5]
```diff
@@ -467,7 +466,7 @@ def objective(trial):
             )
             return
 
-        self.storage.set_trial_intermediate_value(self._trial_id, step, value)
+        self._cached_frozen_trial.intermediate_values[step] = value
```
Comment (c-bata) [inline optuna/trial/_trial.py:469]: You must save the intermediate value to the storage and update the cache here.

```suggestion
        self.storage.set_trial_intermediate_value(self._trial_id, step, value)
        self._cached_frozen_trial.intermediate_values[step] = value
```

Comment (eukaryo) [reply optuna/trial/_trial.py:470]: Thank you very much for the suggestion.

**Comment by c-bata** [comment]
Could you apply the following patch?
https://gist.github.com/c-bata/4287d0f2514e6af03b84acd4293ad8bc

File: optuna/trial/_trial.py [hunk 6]
```diff
@@ -572,21 +573,22 @@ def set_system_attr(self, key: str, value: Any) -> None:
...
         trial_id = self._trial_id
 
-        trial = storage.get_trial(trial_id)
+        trial = self._cached_frozen_trial
 
         if name in trial.distributions:
             # No need to sample if already suggested.
             distributions.check_distribution_compatibility(trial.distributions[name], distribution)
-            param_value = distribution.to_external_repr(storage.get_trial_param(trial_id, name))
+            param_value = distribution.to_external_repr(trial.params[name])
```
Comment (c-bata) [inline optuna/trial/_trial.py:588]:
```suggestion
            param_value = trial.params[name]
```

Comment (eukaryo) [reply optuna/trial/_trial.py:588]: Thank you for the suggestion.

File: optuna/trial/_trial.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (amylase) [reply optuna/trial/_trial.py]: Thank you for the update. looks okay.

**Comment by eukaryo** [comment]
> Could you apply the following patch? https://gist.github.com/c-bata/4287d0f2514e6af03b84acd4293ad8bc

I applied your patch and added type annotation; finally it passed CI.

**Review by amylase — APPROVED** [review]

File: optuna/trial/_trial.py [hunk 7]
```diff
@@ -658,6 +661,11 @@ def _check_distribution(self, name: str, distribution: BaseDistribution) -> None
                 RuntimeWarning,
             )
 
+    def _get_latest_trial(self) -> FrozenTrial:
+        system_attrs = copy.deepcopy(self.storage.get_trial_system_attrs(self._trial_id))
+        self._cached_frozen_trial.system_attrs = system_attrs
+        return self._cached_frozen_trial
```
Comment (c-bata) [inline optuna/trial/_trial.py:667]: Could you add a todo comment?

```suggestion
    def _get_latest_trial(self) -> FrozenTrial:
        # TODO(eukaryo): Remove this method after `system_attrs` property is deprecated.
        system_attrs = copy.deepcopy(self.storage.get_trial_system_attrs(self._trial_id))
        self._cached_frozen_trial.system_attrs = system_attrs
        return self._cached_frozen_trial
```

**Review by c-bata — APPROVED** [review]
Thanks for the update!
LGTM with nits.

File: optuna/trial/_trial.py (hunk 7 above)
Comment (eukaryo) [reply optuna/trial/_trial.py:667]: Agreed.

**Review by c-bata — APPROVED** [review]
